test/fs/test_chmod.c

```diff
@@ -0,0 +1,129 @@
+#include <assert.h>
```

Comment:
Are there not existing tests for chmod?

Reply:
I looked in `test/fs` and didn't see any tests using the `FS.chmod` methods for the JS API, so I added another test just in case. I did find another `test_chmod.c`, but that appeared to be for the C `chmod`. If there are more tests, should I delete my `chmod_test.c`?

Reply:
I general, as we have developed wasmfs, we have tended to add new tests for FS operations that overlap with existing tests.   I'm hoping that over time we can roll this back and have a consistent test of tests for the FS layer that are not specific to wasmfs.

It sounds like we generally lack tests for the FS JS API?   I wonder if we should have a consistent approach to adding JS API tests.  Should we test the JS API for chmod to the existing  test/stat/test_chmod.c?   Or should we add JS API tests in their own files?  Or should the JS API all be tests in one big test "test/fs/test_js_api.c`?

Reply:
> I looked in test/fs and didn't see any tests using the FS.chmod methods for the JS API,

I do see some under another directory:
```
$ grep -r "FS.chmod" test/*
test/unistd/access.c:    FS.writeFile('forbidden', ""); FS.chmod('forbidden', 0o000);
test/unistd/access.c:    FS.writeFile('readable',  ""); FS.chmod('readable',  0o444);
test/unistd/access.c:    FS.writeFile('writeable', ""); FS.chmod('writeable', 0o222);
test/unistd/access.c:    FS.writeFile('allaccess', ""); FS.chmod('allaccess', 0o777);
test/unistd/access.c:    FS.chmod('forbidden', 0o777);
test/unistd/access.c:    FS.chmod('readable',  0o777);
test/unistd/access.c:    FS.chmod('writeable', 0o777);
test/unistd/access.c:    FS.chmod('allaccess', 0o777);
test/unistd/truncate.c:    FS.chmod('toread', 0444);
test/unistd/truncate.c:    FS.chmod('toread', 0o777);
```
It would be good to see if one of those two tests now pass thanks to this PR. If so then just enabling that test would be better (perhaps with some additions there as needed).

If neither of those passes due to other current issues then another option could be to land this without a test, then to work on fixing the other issues blocking that test, then to enable the test when possible. I think that's reasonable in the case that we can see a clear path to getting the test enabled (that is, there isn't a chance we'll just put it off and forget it, and end up with missing testing).

Reply:
I was able to run both tests with WasmFS using `test/runner wasmfs.test_unistd_access wasmfs.test_unistd_truncate` and both tests passed!

Reply:
Oh, and in addition to removing the test in this PR, please enable those two passing tests on continuous integration, by adding them somewhere here:

https://github.com/emscripten-core/emscripten/blob/f016cc00bacf23888580b4485972fdf1ad7f3200/.circleci/config.yml#L517

Reply:
Each test run should always occur in a completely empty/clean directory.  Can you explain a little more the issues you are having?  If there are issues with the runner we should try to fix them rather than work around them.

Reply:
On further investigation I believe I was doing something wrong on my end. I reset changes on each of my branches, cleared the Emscripten cache, and deleted anything left in the `test` directory and everything is passing now. Sorry about the false alarm.

Reply:
Sorry about the repeat notifications, but this issue came back up today and I've isolated the issue to running `test_unistd_access` with `NODEFS` and `NODERAWFS` - `MEMFS` and `WASMFS` pass the test. I think there may be multiple issues going on. 

When I run with only `NODEFS` enabled, I get the `EEXISTS` error. I don't think this is an error with the test runner because only one test is getting run. 

```
/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:130
      throw ex;
      ^
Error
    at new ErrnoError (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:3328:60)
    at Object.symlink (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:2699:17)
    at 67100 (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:1013:20)
    at runEmAsmFunction (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:3898:31)
    at _emscripten_asm_const_int (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:3901:14)
    at <anonymous>:wasm-function[11]:0xc1f
    at main (<anonymous>:wasm-function[12]:0xd72)
    at /home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:723:22
    at callMain (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:4516:15)
    at doRun (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:4566:23)
    at run (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:4581:5)
    at runCaller (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:4501:19)
    at removeRunDependency (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:652:7)
    at receiveInstance (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:864:5)
    at receiveInstantiationResult (/home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:883:5) {
  node: undefined,
  setErrno: [Function (anonymous)],
  errno: 20,
  code: 'EEXIST',
  message: 'File exists'
}
Thrown at:
    at /home/ubuntu/Documents/emsdk/emscripten/out/test/access.js:130:7
    at emit (node:events:369:20)
    at node:internal/process/execution:163:25
    at processPromiseRejections (node:internal/process/promises:245:11)
    at processTicksAndRejections (node:internal/process/task_queues:95:32)
```
![image](https://github.com/emscripten-core/emscripten/assets/55058507/7d48cb3b-356f-4642-b3df-5ddf5a3cbff5)

When I try to run with only `NODERAWFS` enabled, I get an assertion error. Both the symlink and original file get the modified permissions. The tests pass in both MEMFS and WASMFS mode, so could this be an issue with how NODERAWFS handles symlinks?

![image](https://github.com/emscripten-core/emscripten/assets/55058507/6638fc37-d52c-4073-b7d6-f40aea43c1b1)


Finally, I found this old unresolved issues about symlinks in NODEFS (https://github.com/emscripten-core/emscripten/issues/10903). Could these issues be related? Finally, I added the tests to this PR. Would it help to see if the problem is reproducible elsewhere?

Reply:
Testing the code in this PR, I do see that error as well, with or without WasmFS. That is, yes, this seems like it could be a more general issue, maybe related to the last link, or maybe just how Node itself handles symlinks. In any case, I'd say it's low priority to figure that out, since our main goal is to get WasmFS to the same level as the old FS. That is, if something doesn't work in the old FS then I wouldn't worry about WasmFS also failing on it.

Reply:
Should I just remove the lchmod tests from `test_unistd_access` for now? Would it be helpful to restore the original `test_fs_chmod` test that included it?

Reply:
> Should I just remove the lchmod tests from test_unistd_access for now?

Yes, I wouldn't add anything that doesn't pass in either the old FS or WasmFS. (If it passes in one, ifdefing could be an option.)

> Would it be helpful to restore the original test_fs_chmod test that included it?

Sorry, what do you mean by restore? Where was it removed?

Reply:
I removed it in a prior commit because we had `test_unistd_access`, but that test did include passing tests for `lchmod` and `fchmod`.

Reply:
Oh ok, but I guess you restored that test since that commit as I don't see it removed in the current state of this PR? (the only change to anywhere in `test/` seems to be some changes to `test/unistd/access.c`). Anyhow, in general, let's not remove existing tests, and add new passing ones for the new functionality.

Reply:
My mistake, I meant that in a previous commit I had created a new `test_fs_chmod` for this PR that we later decided to remove in favor of using `test/unistd/access.c`. Given the issues with `access.c`, would it be worth bringing this test file back?

Reply:
I'm confused about something else, then. Our goal is to get some testing for the new methods here. Why does it matter whether we add that test code in a new test, or an existing test? That is, I'm not sure why there is a problem with adding to the existing test.

Concretely, I'm suggesting doing something like taking the (I assume) passing test code from your new test from earlier in this PR, and adding that to the existing test, where I hope it would still pass?

Reply:
I think `test_unistd_access` is having some issues when attempting to handle symlinks in `NODEFS` and `NODERAWFS`, so any `symlink` calls would fail and `lchmod` couldn't be called. I was able to successfully move the passing `fchmod` code in to `test_unistd_access` and the tests pass. Since `test_truncate` mainly seems to deal with testing the `truncate` function, I didn't think adding `chmod` tests would be suitable there.

Reply:
I see. Then how about ifdefing so that the symlink parts do not run in `NODE*FS` in `test_unistd_access`?

Reply:
I see the python test code already adds `-DNODEFS` when running in that mode (on this line: `self.emcc_args = orig_compiler_opts + ['-D' + fs]`). But you'd need to add `-DNODERAWFS` too, like this:
```diff
diff --git a/test/test_core.py b/test/test_core.py
index b0afcb192..2857f3d4c 100644
--- a/test/test_core.py
+++ b/test/test_core.py
@@ -6141,7 +6141,7 @@ Module['onRuntimeInitialized'] = function() {
     # Node.js fs.chmod is nearly no-op on Windows
     # TODO: NODERAWFS in WasmFS
     if not WINDOWS and not self.get_setting('WASMFS'):
-      self.emcc_args = orig_compiler_opts
+      self.emcc_args = orig_compiler_opts + ['-DNODERAWFS']
       self.set_setting('NODERAWFS')
       self.do_run_in_out_file_test('unistd/access.c')
 
```

Reply:
Ah, I totally forgot about ifdefing. As a side question, I noticed that the test already called `self.set_setting('NODERAWFS')`. Is this different than modifying `self.emcc_args` and adding `['-DNODERAWFS']`?

Reply:
Yes, those are different `set_setting(..)` sets and Emscripten setting, which ends up as `-sSETTING=VALUE` in the emcc commandline. That affects how Emscripten compiles, in emcc.py and the other compiler code in src/* OTOH, `-DFOO` is a C/C++ compiler define, which only affects the C/C++ preprocessor. That only affects C/C++ source code.

So the former allows you to ifdef in JS library code, while the latter allows you to ifdef in C/C++ code. There are some other differences as well that prevent them from being combined, such as that emscripten settings only apply at the "link" stage , when generating a `.js` and `.wasm` from object files. But the C/C++ preprocessor flags apply at the "compile" stage when converting C/C++ to object files, which is earlier.